File: main112.py

def Solve(used, last, cache):
#  if (used, last) in cache:
#      return cache[(used, last)]
  cur_pos = len(used)
  if cur_pos == len(A):
    return 0

  mn = 1e10
  for idx in range(len(A)):
      if not idx in used:
          next_used = used.union([idx])
          subcost = Solve(next_used, A[idx], cache)
          additional = C[cur_pos] * abs(last - A[idx])
          mn = min(mn, subcost + additional)
  cache[(used, last)] = mn

  #print mn
  return mn

#T = int(raw_input())
#for i in range(T):
#N = int(raw_input())
#A = map(int, raw_input().split())
#C = map(int, raw_input().split())

A = [1,20,21,40]
C = [0,500,1,1]
cache = {}

solutions = [Solve(frozenset([idx]), A[idx], cache) for idx in range(len(A))]
#print solutions

print "solution =", min(solutions)

File: main112.java
import static java.util.Arrays.*;
import static java.lang.Math.*;

import java.io.*;
import java.lang.*;
import java.util.*;

class main112 {
    private int[] A, C;
    private int N;
    private int[] cache;

	public static void main(String[] args) throws IOException
	{
	    new main112().run();
	}
	
	private void run() throws IOException
    {
        InputStreamReader converter = new InputStreamReader(System.in);
        BufferedReader in = new BufferedReader(converter);
        String line = in.readLine();
        Scanner sc = new Scanner(line);
        int numCases = sc.nextInt();
        for (int i = 0; i < numCases; i++) {
            line = in.readLine();
            sc = new Scanner(line);
            N = sc.nextInt();
            line = in.readLine();
            sc = new Scanner(line);
            A = new int[N];
            for (int j = 0; j < N; j++) {
                A[j] = sc.nextInt();
            }
            line = in.readLine();
            sc = new Scanner(line);
            C = new int[N];
            for (int j = 0; j < N; j++) {
                C[j] = sc.nextInt();
            }
			
			cache = new int[1 << (A.length+5)];

            int result = Integer.MAX_VALUE;
            for (int j=0; j<A.length; j++) {
                boolean[] used = new boolean[A.length];
                used[j] = true;
                result = min(result, solve(used, A[j]));
            }

            System.out.println(result);
        }
    }

	private int solve(boolean[] used, int last)
    {
        int numSet = 0;
        int bitset = 0;
        for (int i=0; i<used.length; i++) {
            bitset |= (used[i]?1:0) << (i+4);
            if (used[i])
            {
                numSet++;
            }
        }
        bitset += last;

        if (cache[bitset] > 0)
        {
            return cache[bitset];
        }

        if (numSet == A.length)
            return 0;

        int mn = Integer.MAX_VALUE;
        for (int i=0; i<A.length; i++) {
            if (!used[i])
            {
                boolean[] next_used = new boolean[used.length];
                System.arraycopy(used, 0, next_used, 0, used.length);
                next_used[i] = true;

                int subcost = solve(next_used, A[i]);
                int additional = C[numSet] * abs(last - A[i]);
                mn = min(mn, subcost + additional);
            }
        }

        cache[bitset] = mn;
        return mn;
    }
}
